src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -249,15 +274,102 @@ private void UpdateEndpoints()
-                var segment = newPathSegments[i];
-                for (var j = 0; j < segment.Parts.Count; j++)
+                // Only change optional parameters to required when default values have introduced
+                // endpoints with shortened patterns
+                var updateOptionalParameters = requiredParameterEndpointCount > 0;
+
+                UpdatePathSegments(i, updateOptionalParameters, action, routePattern, newPathSegments, ref allParameterPolicies);
+            }
+
+            if (requiredParameterEndpointRouteOrder != null)
+            {
+                // There are multiple cases where optional parameter parts can come into play:
+                // - /Home/{id?} - single trailing segment optional parameter
+                // - /Home/{id?}/{more?} - multiple trailing segment optional parameter
+                // - /Home.{id?} - optional trailing extension
+                // - /Home/{id?}/{*catchAll} - optional parameter then catch all
+
+                var hasCatchAll = newPathSegments.Last().IsSimple &&
+                    newPathSegments.Last().Parts.Last() is RoutePatternParameterPart parameterPart &&
+                    parameterPart.IsCatchAll;
+                var segmentIndex = hasCatchAll ? newPathSegments.Count - 2 : newPathSegments.Count - 1;
+
+                for (int i = 0; i < requiredParameterEndpointCount; i++)
                 {
-                    var part = segment.Parts[j];
+                    var currentSegmentIndex = segmentIndex - i;
+
+                    var nonOptionalParameterEndpoint = CreateEndpoint(
+                        action,
+                        name,
+                        GetPattern(ref patternStringBuilder, newPathSegments),
+                        newPathSegments,
+                        nonInlineDefaults,
+                        requiredParameterEndpointRouteOrder.Value + i,
+                        dataTokens,
+                        suppressLinkGeneration,
+                        suppressPathMatching);
+                    endpoints.Add(nonOptionalParameterEndpoint);
 
-                    if (part.IsParameter &&
-                        part is RoutePatternParameterPart parameterPart &&
-                        action.RouteValues.ContainsKey(parameterPart.Name))
+                    var segmentParts = newPathSegments[currentSegmentIndex].Parts.ToList();
+                    if (segmentParts.Count == 1)
+                    {
+                        // Don't include optional parameter in final endpoint
+                        newPathSegments = newPathSegments.Take(currentSegmentIndex).ToList();
+                    }
+                    else
+                    {
+                        // Don't include optional extension and separator in final endpoint
+                        newPathSegments[currentSegmentIndex] = RoutePatternFactory.Segment(segmentParts.Take(segmentParts.Count - 2));
```

Comment:
So this seems to apply to complex segments (parts > 1) that contain an optional parameter than than just complex segments with a n optional separator. A complex segment always requires **some** text to be specified so the segment must **always** be present in the URL to match. 

I'm not sure if this is a bug in your code or not, so here's an example.
`/{c=Home}/{a=Index}/{x?}-{y?}`

Basically this always will match/generate a URL with 3 segments. Any example you could write like this where the last segment is a complex segment will never **collapse**. 

It's not super clear to me from reviewing the code if this tries to make that possible.

Reply:
I'll test it and see what happens.

Reply:
It is an invalid route. An optional parameter has to be the last part of a segment. Also an optional parameter can only be proceeded by `.`.

Only valid cases:
`/{x?}/`
`/.{x}/`